### DISCUSSION TITLE
Setup global modifiers for all mocks?

### DISCUSSION THREAD

**Discussion by Geddard** [mswjs/msw#724] [Q&A]
Hi there!

I've been trying to figure out a nice and maintainable way of setting up some global controls for all my mocks.

The first one that can be a good example would be to add a global delay for all responses, or maybe all get responses.

I can't find any case like this anywhere :/ And i don't want to have to go and add `ctx.delay(x)` to each mock handler i created... In my head the ideal would be to create either some global handler that adds the delay to all other mocks at some point in the msw lifecycle, or by adding it to .start() or .use()

But i can't find documentation and can't figure it out myself...

Any ideas on how to achieve this?

**Comment by kettanaito** [comment]
Hey, @Geddard. Thanks for raising this question.

MSW strongly encourages its users to compose, instead of relying on any configuration options. The latter is almost always limiting and won't satisfy each and every use-case, while the composition will. 

When it comes to composing MSW logic, you need to figure out on which level you wish to introduce the customization. In your case, you wish to add a fixed response delay to all response resolvers you write. I'd suggest creating a [custom `response` function](https://mswjs.io/docs/recipes/custom-transformer#defining-a-transformer) that encapsulates your delay requirement:

```js
// ./resWithDelay.js
import { createResponseComposition } from 'msw'

export const resWithDelay = createResponseComposition({
  delay: 500
})
```

Then use the `resWithDelay` instead of the default `res` function in your response resolvers:

```js
// ./handlers.js
import { resWithDelay } from './resWithDelay'

export const handlers = [
  rest.get('/user', (req, _, ctx) => {
    return resWithDelay(ctx.json({ firstName: 'John' }))
  })
]
```

This type of customization scales well and gives you control over the custom logic you wish to have. 

In a similar fashion, you can create custom request handlers and [context utilities](https://mswjs.io/docs/basics/response-transformer#example).

**Reply by Geddard** [reply]
I understand but I still think it would be easier to have a toggle I can put in a modal, that applies a delay based on some input value for example, and configures it on all current responses.
Rather than having to go to a thousand different mocks and wrap them with another function.
If there's no way of doing this then it's a shame and i will have to burn some hours adding this wrapper everywhere i guess 😓

**Reply by kettanaito** [reply]
Well, what you described above is arguably a different use case, although it can still be achieved without resorting to configuration. If you wish _runtime_ to affect your mocking layer, then I suggest encapsulating your abstraction in a custom response resolver. Roughly, like so:

```js
export function withDelay(handler) {
  return (req, res, ctx) => {
    // Grab the custom "?delay=N" query parameter on the page
    // where the request happens.
    const urlParams = new URLSearchParams(location.search)
    const customDelay = urlParams.get('delay')
    const mockedResponse = handler(req, res, ctx)
    
    // Set the mocked response delay if it hasn't been set already.
    mockedResponse.delay = mockedResponse.delay ?? customDelay

   return mockedResponse
  })
}
```
```js
rest.get('/user', withDelay((req, res, ctx) => {
  return res(ctx.json({ firstName: 'John' })
})
```

Yes, it feels like writing your custom setup, because it _is_. It's your setup and you should be in charge of it. Keeping a declarative abstraction that you explicitly use in your handlers is both evident and powerful.

**Reply by kettanaito** [reply]
Configuration is such a last resort that I'd argue about only the most crucial options having the right to be in the configuration (i.e. such as the path to a worker script). Extending configuration "just for this one option" may seem harmless, but it always introduces tenfold more issues than it tries to solve. For example, the worker-level configuration is not reactive, meaning you cannot update it once it's been given to the worker. And that is by design, which we are not going to change.

Adding configuration options is about adding the developer's expectations when all they should expect is that given a request A and a handler A that matches that request, they get the mocked response B. Everything else is an abstraction that you should be in charge of. The maintainer's mission is to establish an API versatile enough so you can build your abstractions, not to hard lock different behaviors behind the configuration options.

**Reply by Geddard** [reply]
It's pretty much the same as before, this would still require me to go to all the configure get handlers and wrap them with something else, which is very inefficient... What I'm asking for is a simple global throttle, like you would apply with the network tab. 
It should be as simple as doing something like:
```
worker.start({
 delay: 500
})
```

**Reply by Geddard** [reply]
> Configuration is such a last resort that I'd argue about only the most crucial options having the right to be in the configuration (i.e. such as the path to a worker script). Extending configuration "just for this one option" may seem harmless, but it always introduces tenfold more issues than it tries to solve. For example, the worker-level configuration is not reactive, meaning you cannot update it once it's been given to the worker. And that is by design, which we are not going to change.
> 
> Adding configuration options is about adding the developer's expectations when all they should expect is that given a request A and a handler A that matches that request, they get the mocked response B. Everything else is an abstraction that you should be in charge of. The maintainer's mission is to establish an API versatile enough so you can build your abstractions, not to hard lock different behaviors behind the configuration options.

It wouldn't be a problem to restart the worker when the delay property changes. 
The idea im trying to introduce is a modal type of component where the testers for example can control which responses they want to have for different endpoints, put some json in a text box, click apply and see this json be the response from now on, this can be easily achieved by restarting teh worker and injecting the new data.
I also would add a toggle to simply turn on/off the worker, or restart it if something is wrong.
I'm only missing delay controls :/

**Reply by Geddard** [reply]
Maybe there's some place where i could do something like 
```
rest.get('/*', (req, res, ctx) => {
      return res(ctx.delay(1000))
    })
```

But im not sure if two handlers can be executed for one get call :( probably not

**Reply by kettanaito** [reply]
> It's pretty much the same as before, this would still require me to go to all the configure get handlers and wrap them with something else, which is very inefficient

I see little inefficiency given you only do this once. The benefits you get from using this logic explicitly, as opposed to providing it as an implicit configuration, outweigh a few minutes you spend on applying such abstraction. 

> It wouldn't be a problem to restart the worker when the delay property changes.

This is exactly the problem: it probably wouldn't. And this initial simplicity may mislead us into thinking it's a good idea. In reality, this would broaden the worker's responsibility, affecting your expectations from the library as a whole. This leads to the code being so options-driven that it stops being predictable. I hope this helps you understand why we're against such global configurations.

Also bear in mind that service workers are quite whimsy when it comes to restarts. Restarting (updating) the worker with the new configuration doesn't guarantee it being applied. Worker's registrations can be cached, which MSW already tries to handle by forcing the worker's un-registration when the page unloads. Even that is too much to be responsible for, but it's tolerable to preserve a good basic experience. In terms of configuration, we wouldn't have to do anything with the worker, it'd likely be a matter of mutating an internal reference to the up-to-date options (most logic happens on the client-side, not in the worker). 

> I also would add a toggle to simply turn on/off the worker, or restart it if something is wrong.

You can start/stop/restart the worker at any time via the API returned from `setupWorker()` call:

```js
window.msw = setupWorker(...handlers)

// Anytime in your app or in the browser's console:
window.msw.stop()
``` 

See all the available API methods under the [setupWorker in the sidebar on the left](https://mswjs.io/docs/api/setup-worker).

> Maybe there's some place where i could do something like

I'm afraid this wildcard handler would match all requests. If any handler returns a response, that response is being used as the mocked response. It's a request-resolver pairing that has no merges of any sort. So the code above would respond and stop looking for any other handlers that might have handled this request (which is correct).

**Reply by kettanaito** [reply]
If I understand your expectations correctly, I think the best solution I can suggest at the moment is to create that mapping between the UI form you give to your testers and the resolver logic in the custom resolver as I've shown above. The workflow may be implemented like this:

1. A user edits the UI form in the modal.
2. Each form input change updates the corresponding URL query parameter. That's also how the options persist.
3. The custom response resolver parametrized the mock's behavior based on the query parameters it parses from the current location. 

I'd like to help you achieve your idea, but I don't find the configuration being a solution.

**Reply by Geddard** [reply]
> If I understand your expectations correctly, I think the best solution I can suggest at the moment is to create that mapping between the UI form you give to your testers and the resolver logic in the custom resolver as I've shown above. The workflow may be implemented like this:
> 
> 1. A user edits the UI form in the modal.
> 2. Each form input change updates the corresponding URL query parameter. That's also how the options persist.
> 3. The custom response resolver parametrized the mock's behavior based on the query parameters it parses from the current location.
> 
> I'd like to help you achieve your idea, but I don't find the configuration being a solution.

I don't really care about this being a config option or something else, but I think having a global delay could be quite useful.
Like I mentioned, something like the throttle that can be configured in the Network tab, or maybe make this throttle affect msw somehow.
But if there simply there is no way now, then I'll have to settle for updating all my handlers at some point to support a configurable delay.

**Reply by Geddard** [reply]
> Well, what you described above is arguably a different use case, although it can still be achieved without resorting to configuration. If you wish _runtime_ to affect your mocking layer, then I suggest encapsulating your abstraction in a custom response resolver. Roughly, like so:
> 
> ```js
> export function withDelay(handler) {
>   return (req, res, ctx) => {
>     // Grab the custom "?delay=N" query parameter on the page
>     // where the request happens.
>     const urlParams = new URLSearchParams(location.search)
>     const customDelay = urlParams.get('delay')
>     const mockedResponse = handler(req, res, ctx)
>     
>     // Set the mocked response delay if it hasn't been set already.
>     mockedResponse.delay = mockedResponse.delay ?? customDelay
> 
>    return mockedResponse
>   })
> }
> ```
> 
> ```js
> rest.get('/user', withDelay((req, res, ctx) => {
>   return res(ctx.json({ firstName: 'John' })
> })
> ```
> 
> Yes, it feels like writing your custom setup, because it _is_. It's your setup and you should be in charge of it. Keeping a declarative abstraction that you explicitly use in your handlers is both evident and powerful.

Hey there again, im trying this approach at the moment, and when i console.log the .delay property in the mocked response, it seems to be there when i execute the call i need but the delay doesn't happen.

Should i take some other step besidesd adding this wrapper?

I tried doing `worker.resetHandlers` with and without params (the params would be the newly formed handlers i think

**Reply by kettanaito** [reply]
> Hey there again, im trying this approach at the moment, and when i console.log the .delay property in the mocked response, it seems to be there when i execute the call i need but the delay doesn't happen.

That's interesting. May I please ask you to set up a reproduction repository for me to look at this? That'd help a lot.

**Reply by Geddard** [reply]
> > Hey there again, im trying this approach at the moment, and when i console.log the .delay property in the mocked response, it seems to be there when i execute the call i need but the delay doesn't happen.
> 
> That's interesting. May I please ask you to set up a reproduction repository for me to look at this? That'd help a lot.

That might be not possible for a long time, could you confirm to me if i should actually resetHandlers from the worker (or server) after updating the global delay value?
Or maybe just hint me to which steps you would take after updating the delay value
